games/xworld3d/tasks/xworld3d_task.py

```diff
@@ -178,6 +182,159 @@ def _parse_collision_event(self, events):
 
         return hits
 
+    def __within_boundary(self, loc):
+        """
+        Determine if a location is out of boundary of the map
+        """
+        h, w = self.env.get_dims()
+        x, y = loc
+        return y >= 0 and y < h and x >= 0 and x < w
+
+    def _get_surrounding_goals(self, distance_threshold=1.5, refer=None):
+        """
+        Given a reference, return all goals within a circular neighborhood
+        with a radius of distance_threshold (excluding the refer itself)
+        Note that distance_threshold is inclusive
+        """
+        goals = self._get_goals()
+        if refer is None:
+            refer, _, _ = self._get_agent()
+            refer = refer.loc
+        ret = []
+        for g in goals:
+            if g.loc == refer:
```

Comment:
In Haichao's dialog task, agent is allowed to be in the same grid of goals. It is better to use id to check duplication.